components/locid_transform/src/canonicalizer.rs

```diff
@@ -266,24 +260,46 @@ impl LocaleCanonicalizer {
     pub const fn new_with_expander(expander: LocaleExpander) -> Self {
         Self {
             aliases: DataPayload::from_static_ref(
-                crate::provider::Baked::SINGLETON_LOCID_TRANSFORM_ALIASES_V1,
+                crate::provider::Baked::SINGLETON_LOCID_TRANSFORM_ALIASES_V2,
             ),
             expander,
         }
     }
 
+    fn try_new_with_expander_compat<P>(
+        provider: &P,
+        expander: LocaleExpander,
+    ) -> Result<Self, LocaleTransformError>
+    where
+        P: DataProvider<AliasesV2Marker> + DataProvider<AliasesV1Marker> + ?Sized,
+    {
+        let payload_v2: Result<DataPayload<AliasesV2Marker>, _> = provider
+            .load(Default::default())
+            .and_then(DataResponse::take_payload);
+        let aliases = if let Ok(payload) = payload_v2 {
+            payload
+        } else {
+            let payload_v1: DataPayload<AliasesV1Marker> = provider
+                .load(Default::default())
+                .and_then(DataResponse::take_payload)?;
+            payload_v1.map_project(|st, _| st.into())
+        };
+
+        Ok(Self { aliases, expander })
+    }
+
     #[doc = icu_provider::gen_any_buffer_unstable_docs!(UNSTABLE, Self::new_with_expander)]
     pub fn try_new_with_expander_unstable<P>(
         provider: &P,
         expander: LocaleExpander,
-    ) -> Result<LocaleCanonicalizer, LocaleTransformError>
+    ) -> Result<Self, LocaleTransformError>
     where
-        P: DataProvider<AliasesV1Marker> + ?Sized,
+        P: DataProvider<AliasesV2Marker> + ?Sized,
     {
-        let aliases: DataPayload<AliasesV1Marker> =
+        let aliases: DataPayload<AliasesV2Marker> =
             provider.load(Default::default())?.take_payload()?;
 
-        Ok(LocaleCanonicalizer { aliases, expander })
+        Ok(Self { aliases, expander })
     }
 
     icu_provider::gen_any_buffer_data_constructors!(
```

Comment:
Issue: you'll need to unpack this macro. It's not too hard though since you can use `icu_provider::gen_any_buffer_unstable_docs!`.

Reply:
Why do I need to unpack it?

Reply:
Code that calls `LocaleCanonicalizer::try_new_with_expander_with_buffer_provider`, they should get the compat constructor, but this macro makes it call the non-compat constructor.